### PR TITLE
Diff.py: Add newlines to outward representation

### LINKED ISSUE #3834: diff: diff.add_lines doesn't append newline to the lines passed as a list
example:
```
diff = Diff([])
diff.add_lines(0, ["1", "2", "3"])
print(diff.unified_diff)
  --- 
  +++ 
  @@ -0,0 +1,3 @@
  +1+2+3

```
while, for `add_lines(line_br, list)` current implementation requires user to append new lines, we should check if the lines in the list endswith a `\n` if not append one implicitly

Opened via [gitter](https://gitter.im/coala/coala/?at=58b8338f1465c46a56d78e16) by @meetmangukiya

### PR DESCRIPTION
**Pull request by mrtes**
Closes #3834 

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

File: coalib/results/Diff.py [hunk 1]
```diff
@@ -286,9 +286,9 @@ def unified_diff(self):
         for equality comparison.
         """
         return ''.join(difflib.unified_diff(
-            self.original,
-            self.modified,
-            tofile=self.rename if isinstance(self.rename, str) else ''))
+                self.original,
+                self.modified,
+                tofile=self.rename if isinstance(self.rename, str) else ''))
```
Comment (Makman2) [inline coalib/results/Diff.py]: is this indentation necessary for this commit? -> New commit.
As it's a codestyle change only, rather omit this change completely as the previous version is also valid.

File: coalib/results/Diff.py [hunk 2]
```diff
@@ -467,7 +467,6 @@ def add_lines(self, line_nr_before, lines):
         """
         if lines == []:
             return  # No action
-
```
Comment (Makman2) [inline coalib/results/Diff.py]: unnecessary codestyle change (and imo makes code less readable too ;D)

File: coalib/results/Diff.py [hunk 3]
```diff
@@ -616,3 +615,30 @@ def remove(self, range):
         :param range: The range to delete.
         """
         self.replace(range, '')
+
+    def _add_linebreaks(self, lines):
+        """
+        Validate that each line in lines end with a
```
Comment (Makman2) [inline coalib/results/Diff.py]: end -> ends

File: coalib/results/Diff.py [hunk 4]
```diff
@@ -616,3 +615,30 @@ def remove(self, range):
...
+        """
+
+        return [line
+                if line.endswith('\n')
+                else line + '\n'
+                for line in lines]
+
+    def _generate_linebreaks(self, lines):
+        """
+        Validate that each line in lines end with a
```
Comment (Makman2) [inline coalib/results/Diff.py]: end -> ends

File: coalib/results/Diff.py [hunk 5]
```diff
@@ -244,10 +243,9 @@ def original(self):
         """
         Retrieves the original file.
         """
-        return self._file
+        return self._generate_linebreaks(self._file)
```
Comment (Makman2) [inline coalib/results/Diff.py]: definitely cache this! This is quite cost intensive if a user needs it several times. Either use `lru_cache(1)` or save it directly on instantiation inside a variable :+1:

File: coalib/results/Diff.py [hunk 6]
```diff
@@ -278,18 +276,37 @@ def modified(self):
...
+        """
+        eof_linebreak = False
+
+        if self._file and self._file[-1].endswith('\n'):
+            eof_linebreak = True
+
+        file = [line.rstrip('\n') for line in self._modified()]
+
+        if eof_linebreak:
+            return self._add_linebreaks(file)
```
Comment (Makman2) [inline coalib/results/Diff.py]: why not using `generate_linebreaks` without r-stripping? `_add_linebreaks` seems to check itself when it has to append a newline^^

File: coalib/results/Diff.py [hunk 7]
```diff
@@ -244,10 +243,9 @@ def original(self):
         """
         Retrieves the original file.
         """
-        return self._file
+        return self._generate_linebreaks(self._file)
 
-    @property
-    def modified(self):
+    def _modified(self):
```
Comment (Makman2) [inline coalib/results/Diff.py]: maybe call it `_raw_modified` or so? It's hard to tell where the difference is between this one and `modified` if you don't know about this `'\n'` conversion thing^^

File: coalib/results/Diff.py [hunk 8]
```diff
@@ -278,18 +276,37 @@ def modified(self):
...
 
         Note that the unified diff is not deterministic and thus not suitable
         for equality comparison.
         """
-        return ''.join(difflib.unified_diff(
-                self.original,
-                self.modified,
+        list_unified_diff = list(difflib.unified_diff(
+                self._file,
+                self._modified(),
```
Comment (Makman2) [inline coalib/results/Diff.py]: why not using `self.original` and `self.modified`? Then you don't have to do `_generate_linebreaks`^^

File: coalib/results/Diff.py [hunk 9]
```diff
@@ -413,7 +430,7 @@ def __add__(self, other):
             if change.add_after is not False:
                 result.add_lines(line_nr, change.add_after)
             if change.change is not False:
-                result.change_line(line_nr, change.change[0], change.change[1])
+                result.modify_line(line_nr, change.change[1])
```
Comment (Makman2) [inline coalib/results/Diff.py]: new commit

File: coalib/results/Diff.py [hunk 10]
```diff
@@ -616,3 +615,30 @@ def remove(self, range):
         :param range: The range to delete.
         """
         self.replace(range, '')
+
+    def _add_linebreaks(self, lines):
```
Comment (Makman2) [inline coalib/results/Diff.py]: can be made a staticmethod

File: coalib/results/Diff.py [hunk 11]
```diff
@@ -616,3 +615,30 @@ def remove(self, range):
...
+
+        :param lines: A list of strings, representing lines.
+        """
+
+        return [line
+                if line.endswith('\n')
+                else line + '\n'
+                for line in lines]
+
+    def _generate_linebreaks(self, lines):
```
Comment (Makman2) [inline coalib/results/Diff.py]: can be made a staticmethod

File: coalib/results/Diff.py (hunk 1 above)
Comment (mrtes) [inline coalib/results/Diff.py]: No, I probably forgot to remove it after trying something. Will remove.

File: coalib/results/Diff.py (hunk 2 above)
Comment (mrtes) [inline coalib/results/Diff.py]: Probably also something I forgot. Normally, I'm all for blank lines :)

File: coalib/results/Diff.py (hunk 5 above)
Comment (mrtes) [inline coalib/results/Diff.py]: You are right. I didn't know about the `lru_cache` thing, seems handy. 
Which approach do you recommend?

File: coalib/results/Diff.py (hunk 6 above)
Comment (mrtes) [inline coalib/results/Diff.py]: Thats a weird edge case with deletions.

Assume a file like this: `file = ['a\n', 'b']`
We delete the last line, resulting in `file = ['a\n']`. 

Since the "eof-state" (i.e. no new line at the end of the file) should be kept, the `\n` now be deleted.
This would require functionality my current functions do not have, and would likely be more complicated to implement:
lots of checks like where did the deletion happen and did we have a `\n` at the eof or not, and probably also some other edge cases. 
Stripping all new lines and then just choosing the correct function, based on the original "eof-state" is far easier. 

Actually, all these edge cases between add, modify and delete led me to just changing the files (or diffs) we hand out from this function, after all changes have been applied.

File: coalib/results/Diff.py (hunk 9 above)
Comment (mrtes) [inline coalib/results/Diff.py]: Should I move all my modifications, that replace `change_line` with `modify_line` into the new commit, or just in functions that i have not altered in other ways?

File: coalib/results/Diff.py (hunk 8 above)
Comment (mrtes) [inline coalib/results/Diff.py]: In a case with "eof-state" without `\n`, this would result in the last line of the original/changed file being on the same line in the diff, something like:
`original = ['1','2','3']`
`target = ['1','2','4']`
`diff = ['+++\n', '---\n', '@@ xxxxx @@\n', '1\n', '2\n', '-3+4']`

I found this weird, because each change should be on a new line.
So I have to run either `_generate_linebreaks` or `_add_linebreaks` anyway to avoid this problem.
(However, the latter is unnecessary, because it doesn't matter if the last line of the diff has a '\n' or not - the representation seems equal).

By using the un-linebreaked files for unified_diff creation and add linebreaks later, I avoid unnecessary calculations.

File: coalib/results/Diff.py (hunk 5 above)
Comment (Makman2) [inline coalib/results/Diff.py]: Up to you ;)

File: coalib/results/Diff.py (hunk 6 above)
Comment (Makman2) [inline coalib/results/Diff.py]: > Since the "eof-state" (i.e. no new line at the end of the file) should be kept, the \n now be deleted.

That's the question, do we wanna keep that? I'm for abandoning the EOF-state, the user needs to change the new last line if he wants to keep the old no-\n-at-EOF-state. As it's in fact a change of that line.

File: coalib/results/Diff.py (hunk 8 above)
Comment (Makman2) [inline coalib/results/Diff.py]: ahhhh I think we still haven't fixed that for unified-diff that no \n at EOF gets handled correctly. There's somewhere an issue and even a PR lying around already for longer time... Could you place an explanatory comment? And maybe a link to the issue if you like :+1:

File: coalib/results/Diff.py (hunk 9 above)
Comment (Makman2) [inline coalib/results/Diff.py]: > Should I move all my modifications, that replace change_line with modify_line into the new commit

:+1:

File: coalib/results/Diff.py (hunk 6 above)
Comment (mrtes) [inline coalib/results/Diff.py]: Well, that's what I understood from the discussion in #3834.

If we do not need to keep the EOF-state, this change will be far less complex because we can just append linebreaks to everything.

File: coalib/results/Diff.py (hunk 8 above)
Comment (mrtes) [inline coalib/results/Diff.py]: Anyway, the `Diff.unified_diff` functions is used nowhere in the core except for the json output.
The printing in `ShowPatchAction` is done by using the difflib version directly. 

Interestingly, I don't see the issue you describe in #2006, even with the unmodified coala version.
Probably due to how printing the diff is handled in `ShowPatchAction.apply`, or rather in the printer, that adds a linebreak anyway ;)

In the end the question is, if the newline stuff in the diff is necessary, or for which usecase.

Comment (Makman2) [inline coalib/results/Diff.py]: if someone wants to check if files do end with \n, we have to support that. Then we have a diff changing the last line from `XYZ` to `XYZ\n`.

**Comment by mrtes** [comment]
Meh, messed up the commits somehow.

**Comment by mrtes** [comment]
I implemented your suggestions @Makman2.
I have kept the EoF-State in modified/original currently, but we can still discuss how to deal with it.

**Comment by mrtes** [comment]
@Makman2  you wanted to be pinged.

File: coalib/results/Diff.py (hunk 6 above)
Comment (Makman2) [inline coalib/results/Diff.py]: > Assume a file like this: file = ['a\n', 'b']
> We delete the last line, resulting in file = ['a\n'].

I just meant this case. When a user supplies a text initially (i.e. via `__init__`) we shall preserve the the EOF-state. Only when we remove that line, we don't preserve state for "new last lines".

File: coalib/results/Diff.py [hunk 12]
```diff
@@ -244,10 +245,9 @@ def original(self):
         """
         Retrieves the original file.
         """
-        return self._file
+        return self._original
 
-    @property
-    def modified(self):
+    def _raw_modified(self):
```
Comment (Makman2) [inline coalib/results/Diff.py]: actually do we still need this property? Maybe we should initially append newlines with `_generate_linebreaks` and consider this the `_original`/`_file` then?

Comment (Makman2) [inline coalib/results/Diff.py]: ugh nvm I thought total nonsense...

Comment (Makman2) [inline coalib/results/Diff.py]: Ah but maybe use in this function instead of `_file` -> `_original`? So then we don't need the real original `_file` any more^^

File: coalib/results/Diff.py [hunk 13]
```diff
@@ -278,18 +278,42 @@ def modified(self):
...
+        file = [line.rstrip('\n') for line in self._raw_modified()]
+
+        if eof_linebreak:
+            return self._add_linebreaks(file)
+        return self._generate_linebreaks(file)
+
+    @property
     def unified_diff(self):
         """
         Generates a unified diff corresponding to this patch.
```
Comment (Makman2) [inline coalib/results/Diff.py]: leave a newline here, as the docs below are detailed documentation :+1:

File: coalib/results/Diff.py (hunk 12 above)
Comment (Makman2) [inline coalib/results/Diff.py]: So we would copy your new `modified` code to `__init__`, initially linebreak-ify the originally and all patches are referring to this version then. We would need to patch `add_line`, `change_line` and so on to properly convert inputs then^^
What do you think?

Comment (mrtes) [inline coalib/results/Diff.py]: I thought keeping `_file` would be a good idea if a un-linebreak-ified version of the file is required somewhere. However, it seems it is not called anywhere outside of Diff, so we could ditch it.

Not sure I understand the other suggestion (or rather, the advantage of it, assuming we keep EOF-state). 
Initially adding the linebreaks will complicate the modification functions, because we have to check for addition/modification/delection locations to preserve the EOF-state.
However, if we ditch that too, and apply linebreaks everywhere (like you suggested earlier) it could be feasible.

We would essentially trade the linebreak-calculations in public data-revealing functions against a one-off linebreak addition on every change.

Comment (Makman2) [inline coalib/results/Diff.py]: Hmm... and now I'm lost on what way to take :D
Let's take yours then, it's definitely a valid approach and your work is then not thrown away just by changing the design ;)

btw: `s/originally/original` (damn sometimes I do some grammatical mistakes... that comes from always changing a sentence :3)

File: coalib/results/Diff.py (hunk 13 above)
Comment (mrtes) [inline coalib/results/Diff.py]: Where should I add the newline?

Comment (mrtes) [inline coalib/results/Diff.py]: Nvm got it.

File: tests/results/DiffTest.py [hunk 14]
```diff
@@ -626,3 +626,21 @@ def test_delete(self):
...
         self.assertEqual(self.uut.modified, [])
         self.uut.delete = False
+
+    def test_linebreaks(self):
+        diff = Diff(['1'])
+        lines = ['1', '2', '3']
+        appended = diff._add_linebreaks(lines)
+        skip_eof = diff._generate_linebreaks(lines)
+        empty = diff._add_linebreaks([])
+        empty_skip = diff._generate_linebreaks([])
```
Comment (Makman2) [inline tests/results/DiffTest.py]: you don't need the diff instance, access the staticmethods from `Diff` instead of `diff` :+1:

File: tests/results/DiffTest.py [hunk 15]
```diff
@@ -626,3 +626,21 @@ def test_delete(self):
...
+
+        for line in skip_eof[:-1]:
+            self.assertRegex(line, '\d\n')
+        self.assertRegex(skip_eof[-1], '\d(?!\n)')
+
+        for line in appended:
+            self.assertRegex(line, '\d\n')
+
+        self.assertEqual(empty, [])
+        self.assertEqual(empty_skip, [])
```
Comment (Makman2) [inline tests/results/DiffTest.py]: you could maybe simplify this a bit by inlining the call made at definition of `empty`. Then those things are closer together which actually belong together ;) That makes it especially easier to follow (as you don't have to think about 4 test variables at the same time)
```python
self.assertEqual(Diff._add_linebreaks([]), [])
self.assertEqual(Diff._add_linebreaks([]), Diff._generate_linebreaks[])
```

File: tests/results/DiffTest.py [hunk 16]
```diff
@@ -626,3 +626,21 @@ def test_delete(self):
         self.uut.delete = True
         self.assertEqual(self.uut.modified, [])
         self.uut.delete = False
+
+    def test_linebreaks(self):
```
Comment (Makman2) [inline tests/results/DiffTest.py]: maybe you want two different test functions for two staticmethods? up to you^^

File: tests/results/DiffTest.py [hunk 17]
```diff
@@ -626,3 +626,21 @@ def test_delete(self):
...
+        diff = Diff(['1'])
+        lines = ['1', '2', '3']
+        appended = diff._add_linebreaks(lines)
+        skip_eof = diff._generate_linebreaks(lines)
+        empty = diff._add_linebreaks([])
+        empty_skip = diff._generate_linebreaks([])
+
+        for line in skip_eof[:-1]:
+            self.assertRegex(line, '\d\n')
```
Comment (Makman2) [inline tests/results/DiffTest.py]: why not asserting against a list? easier code + easier to read + less error prone I think
```
self.assertEqual(Diff._add_linebreaks(lines), ['1\n', '2\n', '3\n'])
```

File: coalib/results/Diff.py (hunk 6 above)
Comment (Makman2) [inline coalib/results/Diff.py]: ping, that still looks like that this will do:
```
return self._add_linebreaks(self._raw_modified)
```

File: tests/results/DiffTest.py (hunk 15 above)
Comment (mrtes) [inline tests/results/DiffTest.py]: Hm, I like to separate Variable initialization from actual use, but I see your point. 

I would still avoid multiple `Diff._add_linebreaks([])` in favor of a variable empty, but maybe directly above the asserts?

File: tests/results/DiffTest.py (hunk 17 above)
Comment (mrtes) [inline tests/results/DiffTest.py]: Great idea, that didn't occur to me until now :+1:

File: tests/results/DiffTest.py (hunk 15 above)
Comment (mrtes) [inline tests/results/DiffTest.py]: Nvm, with the other changes it makes more sense this way.

File: coalib/results/Diff.py [hunk 18]
```diff
@@ -278,18 +278,37 @@ def modified(self):
...
+        return self._generate_linebreaks(self._raw_modified())
+
+    @property
     def unified_diff(self):
         """
         Generates a unified diff corresponding to this patch.
 
+        Each change will be displayed on its own line. Additionally, the
+        unified_diff preserves the EOF-state of the original file. This
+        means that the Diff will only have a linebreak on the last line,
```
Comment (Makman2) [inline coalib/results/Diff.py]: Diff in double backticks so it gets rendered as code ;) Usually code references like these should be enclosed in double-backticks so people know that this is a legal identifier :+1:

File: coalib/results/Diff.py [hunk 19]
```diff
@@ -278,18 +278,37 @@ def modified(self):
...
+        """
+        return self._generate_linebreaks(self._raw_modified())
+
+    @property
     def unified_diff(self):
         """
         Generates a unified diff corresponding to this patch.
 
+        Each change will be displayed on its own line. Additionally, the
+        unified_diff preserves the EOF-state of the original file. This
```
Comment (Makman2) [inline coalib/results/Diff.py]: I think `_` -> `-` is better, unified-diff is a valid term imo :)

File: coalib/results/Diff.py [hunk 20]
```diff
@@ -432,7 +451,7 @@ def __bool__(self):
         """
         return (self.rename is not False or
                 self.delete is True or
-                self.modified != self._file)
+                self._raw_modified() != self._file)
```
Comment (Makman2) [inline coalib/results/Diff.py]: consider this:
```
d = Diff(['1', '2', '3'])
d.modify_line(1, '1\n')
```
This would effectively render no difference in `d.modified` before and after the change, but internally we say that something changed because we handle things without a \n.

File: coalib/results/Diff.py [hunk 21]
```diff
@@ -523,12 +542,12 @@ def modify_line(self, line_nr, replacement):
         if linediff.change is not False and linediff.change[1] != replacement:
             if len(replacement) == len(linediff.change[1]) == 1:
                 raise ConflictError('Cannot merge the given line changes.')
-
+            # strings are used here instead of lists.
```
Comment (Makman2) [inline coalib/results/Diff.py]: don't understand this comment :)

Comment (mrtes) [inline coalib/results/Diff.py]: When a line is modified twice, it is possible for the contents to be merged.
If that is the case, in the lines below 545 new diffs are generated from the changes lines. Since these diffs are not generated from a list, but instead from a string, the elements of these diffs will be chars. 
`diff = Diff('hello')`
 => `diff._file = ['h','e','l','l','o']`
These will be added and formed to a new replacement line. 
Its quite interesting actually, but a bit confusing if you don't know these diffs are created from strings, so I thought I add a comment.

File: coalib/results/Diff.py (hunk 20 above)
Comment (mrtes) [inline coalib/results/Diff.py]: Right.
I changed it to compare `self.modified ` with `self.original`.
Also added a test for this case.

File: coalib/results/Diff.py (hunk 19 above)
Comment (mrtes) [inline coalib/results/Diff.py]: I removed the dash, because it is written "unified diff" in the summary line.

File: coalib/results/Diff.py (hunk 21 above)
Comment (mrtes) [inline coalib/results/Diff.py]: I tried to clarify the comment in the current iteration.

File: coalib/results/Diff.py [hunk 22]
```diff
@@ -523,12 +542,12 @@ def modify_line(self, line_nr, replacement):
         if linediff.change is not False and linediff.change[1] != replacement:
             if len(replacement) == len(linediff.change[1]) == 1:
                 raise ConflictError('Cannot merge the given line changes.')
-
```
Comment (Makman2) [inline coalib/results/Diff.py]: minor and up to you: wanna keep this blank line? I think it improves readability^^

File: coalib/results/Diff.py [hunk 23]
```diff
@@ -278,18 +278,37 @@ def modified(self):
...
         Note that the unified diff is not deterministic and thus not suitable
         for equality comparison.
         """
-        return ''.join(difflib.unified_diff(
-            self.original,
-            self.modified,
+
+        list_unified_diff = list(difflib.unified_diff(
+            self._file,
+            self._raw_modified(),
```
Comment (Makman2) [inline coalib/results/Diff.py]: `_raw_modified` is a property.

File: coalib/results/Diff.py [hunk 24]
```diff
@@ -278,18 +278,37 @@ def modified(self):
...
             tofile=self.rename if isinstance(self.rename, str) else ''))
 
+        return ''.join(self._generate_linebreaks(list_unified_diff))
```
Comment (Makman2) [inline coalib/results/Diff.py]: do we have to do `generate_linebreaks`? As you are using `self._file` and `self._raw_modified` the behaviour should be the same as before, shouldn't it?

File: coalib/results/Diff.py (hunk 23 above)
Comment (mrtes) [inline coalib/results/Diff.py]: Currently `_raw_modified` is implemented as a function.
`modified`, on the other hand is a property.

Anyway, what would be the advantage of using it as a property? 
In this case it should not be used outside the class anyway and does not need to be stored safely as a field, because it has to be recalculated based on possible changes.

File: coalib/results/Diff.py (hunk 24 above)
Comment (mrtes) [inline coalib/results/Diff.py]: Yes, but in case of a file without linebreaks , this would print all changes in one line.
`diff = Diff(['1','2'])`
`diff.modify_line(1,'2')`
`diff.modify_line(2,'1')`
`diff.unified_diff => '--- \n+++ \n@@ -1,2 +1,2 @@\n+2 1-2'`

In a case where only the last line does not have a linebreak it would look even weirder:

`diff = Diff(['1\n','2'])
diff.modify_line(1,'2')
diff.modify_line(2,'1')
diff.unified_diff => '--- \n+++ \n@@ -1,2 +1,2 @@\n-1\n 2+1'`

(I guess the ordering change is due to the unified diff being non-deterministic?)

Comment (mrtes) [inline coalib/results/Diff.py]: The usage of `Diff.from_unified_diff` for a unified diff that is created from an input file without linebreaks is currently broken, too.
(It is just never used)

Comment (Makman2) [inline coalib/results/Diff.py]: Ah so you actually repaired it? Maybe a separate commit for that, explaining the old broken behaviour. You could also file yourself an issue and fix it together :+1:

Comment (Makman2) [inline coalib/results/Diff.py]: (so the variable-change needs to be still here in this commit, but the `return ''.join(self._generate_linebreaks(...)` can be moved out into a new one imo)

File: coalib/results/Diff.py (hunk 23 above)
Comment (Makman2) [inline coalib/results/Diff.py]: oh my bad have misread the diff. Nvm then, a function is totally fine :+1:

Actually there are no differences between functions in properties regarding behaviour, properties just impose different syntax features:
```
a.get_value()
# as property:
a.value

# using setters can be more idiomatic with properties:
a.set_value(3)
a.value = 3
```

File: coalib/results/Diff.py (hunk 24 above)
Comment (mrtes) [inline coalib/results/Diff.py]: Nah, it isn't fixed...yet.
The linebreaks are a first step, but the diff still needs filenames for it to work D:
It would be best if we could access the filename of the original, although a placeholder would work too.
I guess that should really be in a new issue, though.

Still not sure the linebreak addition should be moved though, cause that would require me to change all related tests again ;(

Comment (Makman2) [inline coalib/results/Diff.py]: hmmm... is it that hard? :/
would be nice to have it in a separate commit :/
(`git reset HEAD~1` and `git add -p` are your friends :3)

Comment (mrtes) [inline coalib/results/Diff.py]: No actually it wasn't. Got it mixed up with `from_unified_diff`, which had lots of changes to tests.

File: tests/results/DiffTest.py [hunk 25]
```diff
@@ -626,3 +626,17 @@ def test_delete(self):
...
+            ['1\n', '2\n', '3\n'])
+
+        self.assertEqual(Diff._add_linebreaks([]), [])
+
+    def test_generate_linebreaks(self):
+        self.assertEqual(
+            Diff._generate_linebreaks(['1', '2', '3']),
+            ['1\n', '2\n', '3'])
+
+        self.assertEqual(Diff._generate_linebreaks([]), [])
```
Comment (Makman2) [inline tests/results/DiffTest.py]: ah actually could you add tests when `\n`s are already there? :3

Comment (Makman2) [inline tests/results/DiffTest.py]: so something like `['1\n', '2\n', '3\n']

Comment (Makman2) [inline tests/results/DiffTest.py]: and same for `test_add_linebreaks` :)

Comment (Makman2) [inline tests/results/DiffTest.py]: oh and you may want to test a mixed sample:
```
['1\n', '2', '3\n']
```
or similar

File: coalib/results/Diff.py [hunk 26]
```diff
@@ -278,16 +278,28 @@ def modified(self):
...
         Generates a unified diff corresponding to this patch.
 
         Note that the unified diff is not deterministic and thus not suitable
         for equality comparison.
         """
-        return ''.join(difflib.unified_diff(
-            self.original,
-            self.modified,
+
+        list_unified_diff = list(difflib.unified_diff(
```
Comment (Makman2) [inline coalib/results/Diff.py]: 1. you don't join
2. you don't return it

Tests do pass because they are executed PR wide, not on each commit :/ (would take to long anyway often)

File: coalib/results/Diff.py [hunk 27]
```diff
@@ -432,7 +444,7 @@ def __bool__(self):
         """
         return (self.rename is not False or
                 self.delete is True or
-                self.modified != self._file)
+                self.modified != self.original)
```
Comment (Makman2) [inline coalib/results/Diff.py]: the tests did pass before when comparing `self._raw_modified()` and `self._file`, right? Could you please add a test testing the edge-case I mentioned last time? When modifying a line suddenly with `\n` characters that hadn't some before^^

File: tests/results/ResultTest.py [hunk 28]
```diff
@@ -182,8 +182,13 @@ def test_json_diff(self):
...
+                                                         '-2\n'
+                                                         '-3\n'
+                                                         '+3_changed')
+
         JSONEncoder = create_json_encoder(use_relpath=True)
         json_dump = json.dumps(diff, cls=JSONEncoder, sort_keys=True)
         self.assertEqual(
-            json_dump, '"--- \\n+++ \\n@@ -1,3 +1,2 @@\\n 1-2-3+3_changed"')
+            json_dump,
+            '"--- \\n+++ \\n@@ -1,3 +1,2 @@\\n 1\\n-2\\n-3\\n+3_changed"')
```
Comment (Makman2) [inline tests/results/ResultTest.py]: layout change -> separate commit or remove :)

Comment (Makman2) [inline tests/results/ResultTest.py]: rather remove, the old one is fine imo ;)

**Comment by Makman2** [comment]
You may like to create an issue for dd305f0, but up to you. I think it would be good to track that this is actually a bug that you discovered and fix now :+1:

File: tests/results/ResultTest.py (hunk 28 above)
Comment (mrtes) [inline tests/results/ResultTest.py]: Actually it is a content change (multiple \n), which renders the line longer than allowed.

Comment (Makman2) [inline tests/results/ResultTest.py]: oh my bad :3
(why the heck I haven't seen this... /me is going to buy new eyes...)

File: coalib/results/Diff.py (hunk 26 above)
Comment (Makman2) [inline coalib/results/Diff.py]: ping

Comment (Makman2) [inline coalib/results/Diff.py]: oops no bump, change was made below.

File: coalib/results/Diff.py [hunk 29]
```diff
@@ -278,18 +278,32 @@ def modified(self):
...
-        return ''.join(difflib.unified_diff(
-            self.original,
-            self.modified,
+
+        list_unified_diff = list(difflib.unified_diff(
+            self._file,
+            self._raw_modified(),
             tofile=self.rename if isinstance(self.rename, str) else ''))
 
+        return ''.join(list_unified_diff)
```
Comment (Makman2) [inline coalib/results/Diff.py]: I would prefer if you keep the old code structure and just change the variables. Easier to understand for foreigners + you currently create a list of the diff, which is not necessary. Keeping the old structure automatically avoids this :+1:

File: tests/results/DiffTest.py [hunk 30]
```diff
@@ -628,15 +630,30 @@ def test_delete(self):
...
 
         self.assertEqual(Diff._generate_linebreaks([]), [])
+
+        self.assertEqual(
+            Diff._generate_linebreaks(no_lb_eof),
+            no_lb_eof)
+
+        self.assertEqual(
+            Diff._generate_linebreaks(lb_eof),
+            lb_eof)
```
Comment (Makman2) [inline tests/results/DiffTest.py]: I think those changes should go into the other commit introducing `generate_linebreaks` and `add_linebreaks`

(GitHub does display the same comments in different commits if they touch the same parts of code, sadly... not really fitting our review style^^)

File: coalib/results/Diff.py [hunk 31]
```diff
@@ -278,16 +278,28 @@ def modified(self):
...
+
+    @property
     def unified_diff(self):
         """
         Generates a unified diff corresponding to this patch.
 
         Note that the unified diff is not deterministic and thus not suitable
         for equality comparison.
         """
+
```
Comment (Makman2) [inline coalib/results/Diff.py]: hmm let's be conform with everything else ;) so don't remove the newline.

Main problem is that those style changes are likely reverted again by someone else who's modifying this code. So we get many commits adding a newline, removing it again, adding it, and so on and so on.

Comment (mrtes) [inline coalib/results/Diff.py]: So should I remove it, or not?

Comment (Makman2) [inline coalib/results/Diff.py]: ah egh sorry I meant to remove it :3

Comment (Makman2) [inline coalib/results/Diff.py]: nvm :)

File: tests/results/DiffTest.py [hunk 32]
```diff
@@ -138,10 +138,10 @@ def test_modified(self):
         self.uut.change_line(3, '3', '3.changed')
 
         self.assertEqual(self.uut.modified, result_file)
-        self.assertEqual(self.uut.original, self.file)
 
         self.uut.delete_line(len(self.file))
-        del result_file[len(result_file) - 1]
+        del result_file[-1]
+        result_file[-1] = result_file[-1].rstrip('\n')
```
Comment (Makman2) [inline tests/results/DiffTest.py]: ah this is quite an edge case ... if the old version has no \n's and we delete the last line... though not really important for now, we should maybe document that case more properly later or change behaviour, don't know ;)

**Comment by Makman2** [comment]
ack 0e8b8d5 34a4520 16b3938

**Comment by Makman2** [comment]
@rultor merge
